### PR TITLE
Replace all_gather in device mesh with functional collective equivalent

### PR DESCRIPTION
**Pull request by voznesenskym**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #104056
* #104447

### CODE REVIEW

**Review by wconstab — COMMENTED** [review]
i think this looks correct. i'll let @wanchaol stamp

File: torch/distributed/_tensor/device_mesh.py [hunk 1]
```diff
@@ -172,8 +171,12 @@ def _init_process_groups(self):
         # validate that all calling ranks pass in the same `mesh` argument.
         self_mesh = self.mesh.to(self.device_type)
         mesh_list = [self_mesh.clone() for _ in range(get_world_size())]
-        all_gather(mesh_list, self_mesh)
-        for other_rank, other_mesh in enumerate(mesh_list):
+        mesh_tensor = torch.cat(mesh_list)
+        funcol.all_gather_tensor(mesh_tensor, gather_dim=0, group=_get_default_group())
```
Comment (kumpera) [inline torch/distributed/_tensor/device_mesh.py:175]: you must to use the result of `all_gather_tensor` as functional collectives are not in-place.

**Review by kumpera — CHANGES_REQUESTED** [review]

File: torch/distributed/_tensor/device_mesh.py (hunk 1 above)
Comment (wconstab) [reply torch/distributed/_tensor/device_mesh.py]: yes good catch. thats hard to remember from looking at the apis :/ can we improve somehow?

**Review by kumpera — APPROVED** [review]
Looks good! Thanks for the fix.

File: torch/distributed/_tensor/device_mesh.py [hunk 2]
```diff
@@ -172,8 +171,12 @@ def _init_process_groups(self):
         # validate that all calling ranks pass in the same `mesh` argument.
         self_mesh = self.mesh.to(self.device_type)
         mesh_list = [self_mesh.clone() for _ in range(get_world_size())]
-        all_gather(mesh_list, self_mesh)
-        for other_rank, other_mesh in enumerate(mesh_list):
+        mesh_tensor = torch.cat(mesh_list)
```
Comment (wanchaol) [inline torch/distributed/_tensor/device_mesh.py:174]: nit: if uses functional collective, we don't need the `mesh_list` and the `mesh_tensor` cat, we can just do `funcol.all_gather_tensor(self_mesh, gather_dim=0)` instead.

Comment (voznesenskym) [reply torch/distributed/_tensor/device_mesh.py:174]: sg, ty :)

**Review by wanchaol — APPROVED** [review]

**Comment by wconstab** [comment]
@kumpera for this test error

```
RuntimeError: DeviceMesh initialization does not allow different mesh argument:rank 0 has mesh tensor([0, 1, 2, 3], devic
e='cuda:0') while rank 0has mesh tensor([0, 0, 0, 0], device='cuda:0')!
```

the test case seems to use fake-pg, which means all-gather will be a no-op, so it is totally expected that `mesh_tensor` returns all zeros from each rank.

In the previous case it looks like we snuck by this by putting the 'right answer' into the output tensor for allgather and then relying on allgather being a no-op, so the following assert would pass based on the correct input we put there.

should we (a) run this test under real PG instead, (b) relax the assertion under the fake-pg case (and perhaps 'fix' the mesh tensor) , or something else?

File: torch/distributed/_tensor/device_mesh.py (hunk 1 above)
Comment (kumpera) [reply torch/distributed/_tensor/device_mesh.py]: We plan to remove the `_into_tensor` and `_tensor` suffixes that we think are confusing.

**Comment by kumpera** [comment]
> should we (a) run this test under real PG instead, (b) relax the assertion under the fake-pg case (and perhaps 'fix' the mesh tensor) , or something else?

The simplest solution is to use MTPG as the backend instead of FakePG. FakePG should not be used if there's data depend behavior involved.

Some dtensor methods have a parameter to disable consistency checks like here: https://github.com/pytorch/pytorch/blob/main/torch/distributed/_tensor/api.py#L240C9-L240C18

This is an alternative but I'd say that for this case it doesn't make sense since we want to verify that the check is correctly implemented.

**Comment by wconstab** [comment]
ok cool- i think using MTPG sounds like the best solution then. i'll take a stab at it

**Comment by wanchaol** [comment]
Another simple fix is to make the fake_pg capable of doing some simple validation logic in allgather, https://github.com/pytorch/pytorch/pull/104213

Agreed we can also rely on MTPG for correctness, but we would need a way to make device mesh be constructed under fake pg so that we can initialize the device mesh in a single device/rank.

**Comment by wconstab** [comment]
> Another simple fix is to make the fake_pg capable of doing some simple validation logic in allgather, https://github.com/pytorch/pytorch/pull/104213

i'm good with this approach. What swayed me from using MTPG in this test case was that we literally want to support DTensor+FakePG for other use cases outside of this test failure, so it comes down to either making FakePG work for DTensor init or making DTensor init special-case its safety check for FakePG

**Pull request closed without merging**